stats/kafka/collector.go

```diff
@@ -0,0 +1,125 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2016 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package kafka
+
+import (
+	"context"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/loadimpact/k6/lib"
+	"github.com/loadimpact/k6/stats"
+	log "github.com/sirupsen/logrus"
+)
+
+// Collector implements the lib.Collector interface and should be used only for testing
+type Collector struct {
+	Producer sarama.SyncProducer
+	Config   Config
+
+	Samples []stats.Sample
+}
+
+// New creates an instance of the collector
+func New(conf Config) (*Collector, error) {
+	producer, err := sarama.NewSyncProducer(StringsFromNulls(conf.Brokers), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Collector{
+		Producer: producer,
+		Config:   conf,
+	}, nil
+}
+
+// Init does nothing, it's only included to satisfy the lib.Collector interface
+func (c *Collector) Init() error { return nil }
+
+// Run just blocks until the context is done
+func (c *Collector) Run(ctx context.Context) {
+	log.WithFields(log.Fields{
+		"brokers": c.Config.Brokers,
+		"topic":   c.Config.Topic,
+	}).Debug("Kafka: Sending metrics.")
+
+	<-ctx.Done()
+
+	log.Debug("Kafka: Done sending metrics.")
+
+	err := c.Producer.Close()
+	if err != nil {
+		log.WithError(err).Warn("Kafka: Failed to close producer.")
+	}
+
+	log.Debug("Kafka: Producer closed.")
+}
+
+// Collect just appends all of the samples passed to it to the internal sample slice.
+// According to the the lib.Collector interface, it should never be called concurrently,
+// so there's no locking on purpose - that way Go's race condition detector can actually
+// detect incorrect usage.
+// Also, theoretically the collector doesn't have to actually Run() before samples start
+// being collected, it only has to be initialized.
+func (c *Collector) Collect(scs []stats.SampleContainer) {
```

Comment:
The comment here also looks copy-pasted.

Also, why did you remove the buffering? Generally the idea is for the `Collect()` function to only collect and buffer samples (with minimal processing if needed), not to actually send them. At least not if the sending is actually synchronous, and it looks like  `c.Producer.SendMessage()` is. That's because calling `Collect()` [blocks the engine](https://github.com/loadimpact/k6/blob/master/core/engine.go) and would deteriorate the performance of everything if we're waiting to send stuff over the network, especially once we implement [this](https://github.com/loadimpact/k6/issues/614).
Also, buffering samples would lead to better batching and fewer network requests in total.

Reply:
I can look into this more.  When I was working on running multiple collectors there were panics coming from sending from `Run()` but when I moved it to collect (after seeing that is where `json` was writing the file) I had no issues.

Reply:
Yeah, but the `json` generally won't have to wait for network requests. And it's mostly meant for debugging, not for actual production usage.

Reply:
So I put this back to how it was before.  Will test some more to see if it has issues again with multiple collectors or not now.

Reply:
Hmm seems like I can hide only an individual comment but not the whole inline thread... Anyway, this has been resolved, though the method comment could probably be shortened, since most of the stuff after the first line is useful only for the `dummy` collector and how it's used in the automated tests.